lib/logger.ts

```diff
@@ -23,41 +23,43 @@
 // POSSIBILITY OF SUCH DAMAGE.
 
 import os from 'os';
+import {Writable} from 'stream';
 
 import {LEVEL, MESSAGE} from 'triple-beam';
 import winston from 'winston';
 import LokiTransport from 'winston-loki';
 import {Papertrail} from 'winston-papertrail';
-import TransportStream from 'winston-transport';
+import TransportStream, {TransportStreamOptions} from 'winston-transport';
 
 const consoleTransportInstance = new winston.transports.Console();
 export const logger = winston.createLogger({
     format: winston.format.combine(winston.format.colorize(), winston.format.splat(), winston.format.simple()),
     transports: [consoleTransportInstance],
 });
 
-logger.stream = {
-    write: message => {
-        logger.info(message.trim());
-    },
-};
-
-logger.warnStream = {
-    write: message => {
-        logger.warn(message.trim());
-    },
-};
-
-logger.errStream = {
-    write: message => {
-        logger.error(message.trim());
-    },
-};
-
-// Our own transport which uses Papertrail under the hood but better adapts it to work
-// in winston 3.0
+export function makeLogStream(level: string, logger_: winston.Logger = logger): {write: (string) => void} {
```

Comment:
where did you get this magic code? i'd like to understand the logic of this

Reply:
From https://nodejs.org/api/stream.html#simplified-construction (if you're asking how to make a custom stream?)

Reply:
It replaces the sort-of-vague "we know that morgan only calls `stream.write()` with complete messages so let's abuse that". This can now support anything (morgan expects a real stream, defaulting to "process.stdout" and I wanted to learn how to do this anyway) :)

Reply:
but we know it works as is, so my question would be; what does this actually solve?

Reply:
but we know it works as is, so my question would be; what does this actually solve?

Reply:
It solves the "why lie to the type system when you can actually use the right class"

Reply:
if the buffer does not have a \n, things arent passed to the `logger_.log()`, is that how it works now?

Reply:
Ope good catch

Reply:
It's not how it works now, but...
- morgan expects a "stream"
- morgan writes (arbitrary) line-separated data. Currently in exactly one call to write() (though this isn't guaranteed)
- we (ab)used this to expect every call to write = one log message. And then we stripped off the `\n` if present

My take was, morgan _has_ to write a newline at the end of anything it cares about, because it's writing to a line-buffered output (in all likelihood). Nothing in the contract between morgan and the "stream" says it has to write exactly one message per call to write though. So it seems to me we should act like a line-buffered output, where every time a newline comes along we flush whatever we had.

Otherwise if the morgan folks update to do stuff like `stream.write("moo"); stream.write("\n");` then we'll send two log messages. Which doesn't seem right.

I was trying to do the right thing here, but this has caused a ton of back and forth. I'm happy to revert back to relying on the old way and hope morgan doesn't change -- I mean, it hasn't so far in all the times we've upgraded -- and then just ignore any typing on it.

Reply:
I get what you're saying. I'll give it some more thought. (maybe we need a max buffer size and/or signal catch -> dump buffer kind of thing just to make sure things dont get lost)

Reply:
Thanks for the explanation, makes sense. I think it would be helpful to document this subtlety in the code (and that way when things break we have a little easier time finding it 😛)

Reply:
> this subtlety

Which bit? :) the "if morgan doesn't send newlines, we break"? That seems like "if there's no newlines, you don't get lines of text output", which I can say with something like:

```ts
// we log each complete line output to the stream
```

Something like that? :) Will hang on @partouf too, no hurry on this. And happy to break into a separate PR.

Reply:
I guess an alternative is to _assert_ that everything ends with exactly one newline: then we get the old behaviour exactly (no buffering etc) PLUS a useful error if morgan ever changes.

Reply:
Yeah I was mainly thinking it'd be nice to document the assumption that morgan always writes newline-terminated messages

Reply:
I see. But... if it was a terminal, or file (which morgan is supposedly writing to) would you imagine you'd document "Writes newline separated log lines" ? :)

Reply:
I think I'm just being cantankerous now :D

Reply:
> I see. But... if it was a terminal, or file (which morgan is supposedly writing to) would you imagine you'd document "Writes newline separated log lines" ? :)

Fair enough, you make a good point. But thank you for adding the comment anyway!